classes/privacy/provider.php

```diff
@@ -25,19 +25,240 @@
 defined('MOODLE_INTERNAL') || die;
 use core_privacy\local\metadata\null_provider;
 use core_privacy\local\legacy_polyfill;
+use core_privacy\local\metadata\collection;
+use core_privacy\local\request\contextlist;
+use core_privacy\local\request\approved_contextlist;
+use core_privacy\local\request\approved_userlist;
+use core_privacy\local\request\writer;
+use core_privacy\local\request\userlist;
+
 /**
  * Class provider
  * @package auth_enrolkey\privacy
  */
-class provider implements null_provider {
-    use legacy_polyfill;
+class provider implements \core_privacy\local\metadata\provider,
+                          \core_privacy\local\request\core_userlist_provider,
+                          \core_privacy\local\request\plugin\provider
+{
     /**
-     * Get the language string identifier with the component's language
-     * file to explain why this plugin stores no data.
+     * Returns metadata about this plugin's privacy policy.
      *
-     * @return  string
+     * @param collection $collection The initialised collection to add items to.
+     * @return  collection     A listing of user data stored through this system.
      */
-    public static function _get_reason() {
-        return 'privacy:metadata';
+    public static function get_metadata(collection $collection): collection {
+        $collection->add_database_table(
+                'auth_enrolkey_redirect',
+                [
+                        'usermodified' => 'privacy:metadata:auth_enrolkey_redirect:usermodified',
+                ],
+                'privacy:metadata:auth_enrolkey_redirect'
+        );
+        $collection->add_database_table(
+                'auth_enrolkey_profile',
+                [
+                        'usermodified' => 'privacy:metadata:auth_enrolkey_profile:usermodified',
+                ],
+                'privacy:metadata:auth_enrolkey_profile'
+        );
+        $collection->add_database_table(
+                'auth_enrolkey_cohort',
+                [
+                        'usermodified' => 'privacy:metadata:auth_enrolkey_cohort:usermodified',
+                ],
+                'privacy:metadata:auth_enrolkey_cohort'
+        );
+
+        return $collection;
     }
-}
\ No newline at end of file
+
+    /**
+     * Get the list of contexts that contain user information for the given user.
+     *
+     * @param int $userid the userid to search.
+     * @return contextlist the contexts in which data is contained.
+     */
+    public static function get_contexts_for_userid(int $userid): contextlist {
+        $sql = "SELECT ctx.id
+                  FROM {user_enrolments} ue
+                  JOIN {enrol} e
```

Comment:
I don't believe we need the enrol table as this is not an enrol plugin.

Maybe use something like 
```
        $sql = "SELECT ctx.id
                  FROM {auth_enrolkey_redirect} aer
                  JOIN {context} ctx ON ctx.instanceid = aer.usermodified AND ctx.contextlevel = :contextlevel
                 WHERE aer.usermodified = :userid";
        $params = ['userid' => $userid, 'contextlevel' => CONTEXT_USER];
        $contextlist = new contextlist();
        $contextlist->add_from_sql($sql, $params);

        return $contextlist;
```